ci: Upgrade Android SDK/NDK and refactor to use sdkmanager/avdmanager.

kennytm opened the pull request:
* SDK tools is upgraded to 27.0.0. 

* Stopped using the deprecated `android` tool, instead use the recommended `sdkmanager` and `avdmanager`.

* NDK is upgrade to r15c.

    The r15 series [dropped support for android-9](https://github.com/android-ndk/ndk/wiki/Changelog-r15) (2.3 / Gingerbread), the minimal supported version is now android-14 (4.0 / Ice Cream Sandwich).

r? @alexcrichton 

(WIP, haven't confirmed if the change really works yet)

alexcrichton commented:
Looks great to me, thanks! r=me with a passing set of CI

malbarbo commented:
@kennytm I'm working on libc android ci, see https://github.com/rust-lang/libc/pull/825. It's is necessary to pass `--no_https`.

kennytm commented:
@malbarbo Thanks. This was #45193. I'll see if we could avoid passing `--no_https`...

kennytm commented:
So this is the final list of changes and rationale:

<dl>
<dt>1. Updated Android SDK tools to 27.0.0</dt>
<dd>We saw Android CI failing in liblibc (rust-lang/libc#825)</dd>
<dt>2. Updated NDK to r15c</dt>
<dd>Because why not</dd>
<dt>3. Replaced <code>android</code> by <code>sdkmanager</code> and <code>avdmanager</code></dt>
<dd><code>android</code> was deprecated, which we thought may be another cause of liblibc failing.</dd>
<dt>4. Raised minimum API level of 32-bit targets from 9 (2.3) to 14 (4.0)</dt>
<dd>NDK r15c dropped support for android-9. (The minimum API level for 64-bit targets remains at 21 (5.0))</dd>
<dt>5. Downgraded the Java version used for Android SDK to <code>openjdk-8-jre-headless</code></dt>
<dd>To use HTTPS again (i.e. revert #45193).</dd>
<dt>6. Switched the default compiler for Android from GCC to clang</dt>
<dd>GCC is deprecated on r15c. There is <a href="https://travis-ci.org/rust-lang/rust/builds/294119281">linker error</a> if we don't use clang.</dd>
<dt>7. Switched the C++ standard library from <code>gnustl</code> to <code>libc++</code></dt>
<dd>To match the compiler (but I guess we didn't really use the C++ part of the NDK).</dd>
</dl>

(The last 3 changes are new compared with the first revision)

alexcrichton commented:
All looks and sounds great to me, thanks so much for investigating all this @kennytm!

kennytm commented:
@bors r=alexcrichton

bors commented:
:pushpin: Commit c46b04c has been approved by `alexcrichton`

<!-- @bors r=alexcrichton c46b04cbdd1bdf0183b121fac18513c0609962ee -->

bors commented:
:hourglass: Testing commit c46b04cbdd1bdf0183b121fac18513c0609962ee with merge 49edcbd32ad044c626b66ae7645bd31d7c71786b...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.5146), [status-travis](https://travis-ci.org/rust-lang/rust/builds/294279402?utm_source=github_status&utm_medium=notification)
Approved by: alexcrichton
Pushing 49edcbd32ad044c626b66ae7645bd31d7c71786b to master...
